### PR TITLE
break: fetch new snapshots

### PR DESCRIPTION
**Pull request by pedrotambo**
This PR adds the functionality of reading snapshots of the new endpoint `/snapshots` that provides the active entities in multiple snapshots instead of a whole big one.
A `synchronizer` was added that is in charge of syncing the specified servers by deploying all the entities from snapshots and pointer-changes. It has mechanisms for retrying when things fail.
Also, with the possibility that a snapshot converges, if there are two servers (or more) with the same snapshot, it is downloaded only from one of them, and the entities inside are fetched from all the servers that have that snapshot as a load balancing mechanism.
The `synchronizer` maintains the servers in 3 states:
1. Bootstrap from snapshots: deploying the entities in the snapshots of that server.
2. Bootstrap from pointer-changes: deploying entities from pointer-changes between the last timestamp of its snapshots and now.
3. Syncing: deploying entities polling from pointer-changes.
When a server finishes the tasks in one state, goes to the next one. Retry mechanisms are done if something fails.

When all the initial servers to sync try to perform at least once with states (1) and (2), the initial bootstrap is considered as done, and the Catalysts startup. Note: if some server fails, it'll retry later but the initial bootstrap will finishes anyway.

### CODE REVIEW

File: src/file-processor.ts [hunk 1]
```diff
@@ -40,14 +41,26 @@ export async function* processDeploymentsInFile(
...
     const theLine = line.trim()
     if (theLine.startsWith('{') && theLine.endsWith('}')) {
-      const deployment = coerceEntityDeployment(JSON.parse(theLine))
-      if (deployment) {
-        yield deployment
+      const parsedLine = JSON.parse(theLine)
+      if (SnapshotSyncDeployment.validate(parsedLine)) {
+        yield parsedLine
+      } else if (PointerChangesSyncDeployment.validate(parsedLine)) {
+        yield parsedLine
```
Comment (agusaldasoro) [inline src/file-processor.ts:54]:
```suggestion
      if (SnapshotSyncDeployment.validate(parsedLine) || PointerChangesSyncDeployment.validate(parsedLine)) {
        yield parsedLine
```

File: src/file-processor.ts [hunk 2]
```diff
@@ -40,14 +41,26 @@ export async function* processDeploymentsInFile(
...
+      } else {
+        const errors = {
+          ...SnapshotSyncDeployment.validate.errors,
```
Comment (agusaldasoro) [inline src/file-processor.ts:57]: are those global vars or something like that?

File: src/file-processor.ts [hunk 3]
```diff
@@ -40,14 +41,26 @@ export async function* processDeploymentsInFile(
...
+): AsyncIterable<SyncDeployment> {
   for await (const line of processLineByLine(stream)) {
     const theLine = line.trim()
     if (theLine.startsWith('{') && theLine.endsWith('}')) {
-      const deployment = coerceEntityDeployment(JSON.parse(theLine))
-      if (deployment) {
-        yield deployment
+      const parsedLine = JSON.parse(theLine)
+      if (SnapshotSyncDeployment.validate(parsedLine)) {
+        yield parsedLine
```
Comment (agusaldasoro) [inline src/file-processor.ts:52]: what about the coercion here? Is it no longer necessary?

File: src/index.ts [hunk 4]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
- * @public
+ * @internal
  */
-export async function* getDeployedEntitiesStream(
-  components: SnapshotsFetcherComponents,
-  options: DeployedEntityStreamOptions
-): AsyncIterable<DeploymentWithAuthChain & { snapshotHash?: string }> {
-  const logs = components.logs.getLogger(`getDeployedEntitiesStream(${options.contentServer})`)
-  // the minimum timestamp we are looking for
+export async function* getDeployedEntitiesStreamFromSnapshots(
```
Comment (agusaldasoro) [inline src/index.ts:161]: is it okay to export it then?

File: src/index.ts [hunk 5]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
+      const snapshotInfo = snapshotInfoByHash.get(snapshot.hash)
 
-  // the greatest timestamp we processed
-  let greatestProcessedTimestamp = genesisTimestamp
+      const greatestEndTimestamp = snapshotInfo
+        ? Math.max(snapshotInfo.greatestEndTimestamp, snapshot.lastIncludedDeploymentTimestamp)
+        : snapshot.lastIncludedDeploymentTimestamp
 
-  const metricLabels = contentServerMetricLabels(options.contentServer)
+      const replacedSnapshotHashes = snapshotInfo?.replacedSnapshotHashes ?? []
```
Comment (agusaldasoro) [inline src/index.ts:188]: will this array converge for all catalysts too?

File: src/index.ts [hunk 6]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
+      if (snapshot.replacedSnapshotHashes) {
+        replacedSnapshotHashes.push(snapshot.replacedSnapshotHashes)
```
Comment (agusaldasoro) [inline src/index.ts:190]: will this have repeated items?

File: src/index.ts [hunk 7]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
-      hash &&
-      lastIncludedDeploymentTimestamp &&
-      lastIncludedDeploymentTimestamp > genesisTimestamp &&
-      await components.processedSnapshots.shouldStream(hash, replacedSnapshotHashes)
+      greatestEndTimestamp > genesisTimestamp &&
+      await components.processedSnapshots.shouldProcessSnapshot(snapshotHash, replacedSnapshotHashes)
 
     if (shouldStreamSnapshot) {
       try {
         // 2.1. download the snapshot file if needed
```
Comment (agusaldasoro) [inline src/index.ts:211]: is the number of this comment still valid? As some comments were deleted

File: src/index.ts [hunk 8]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
         let numberOfStreamedEntities = 0
-        for await (const rawDeployment of deploymentsInFile) {
-          const deployment = coerceEntityDeployment(rawDeployment)
-          if (!deployment) continue
-          // selectively ignore deployments by localTimestamp
-          if (deployment.localTimestamp >= genesisTimestamp) {
-            components.metrics.increment('dcl_entities_deployments_processed_total', metricLabels)
+        for await (const deployment of deploymentsInFile) {
+
+          const deploymentTimestamp = 'entityTimestamp' in deployment ? deployment.entityTimestamp : deployment.localTimestamp
```
Comment (agusaldasoro) [inline src/index.ts:228]: what about coerce here too?

File: src/index.ts [hunk 9]
```diff
@@ -150,112 +154,134 @@ export async function downloadEntityAndContentFiles(
...
+  // fetch the /pointer-changes of the remote server using the last timestamp from the previous step with a grace period of 20 min
+  const genesisTimestamp = options.fromTimestamp || 0
+  let greatestLocalTimestampProcessed = genesisTimestamp
   do {
-    // 3.1. download pointer changes and yield
-    const pointerChanges = fetchPointerChanges(components, options.contentServer, greatestProcessedTimestamp)
-    for await (const rawDeployment of pointerChanges) {
-      const deployment = coerceEntityDeployment(rawDeployment)
-      if (!deployment) continue
+    // 1. download pointer changes and yield
```
Comment (agusaldasoro) [inline src/index.ts:269]:
```suggestion
    // download pointer changes and yield
```
?

File: src/processed-snapshots.ts [hunk 10]
```diff
@@ -2,8 +2,23 @@ import { IProcessedSnapshotsComponent, SnapshotsFetcherComponents } from "./type
...
+ * Each entity streamed within a snapshot will be processed by the {@link catalyst#batchDeployer}. It adds each one to a queue
+ * and they are processed in background. Once a deploy in background finishes, it signals this component that one
+ * entity from the snapshot was processed using {@link IProcessedSnapshotsComponent#entityProcessedFrom}.
+ *
+ * When the stream of all deployments it's done within a snapshot, then the signal {@link IProcessedSnapshotsComponent#endStreamOf} is called together with
+ * the quantity of entities streamed. If the quantity matches the number of {@link IProcessedSnapshotsComponent#entityProcessedFrom} signals, it saves
+ * the snapshot as processed. If not, every time the signal {@link IProcessedSnapshotsComponent#entityProcessedFrom} is called, it will check if this
+ * condition is met in order to save or not the snapshot as procesed.
+ *
+ *
```
Comment (agusaldasoro) [inline src/processed-snapshots.ts:20]: 👏

File: src/synchronizer.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+          const lastEntityTimestamp = lastEntityTimestampFromSnapshotsByServer.get(bootstrappingServersFromPointerChange)
+          if (lastEntityTimestamp == undefined) {
+            throw new Error(`Can't start pointer changes stream without last entity timestamp for ${bootstrappingServersFromPointerChange}. This should never happen.`)
+          }
+          const fromTimestamp = lastEntityTimestamp - 20 * 60_000
+          await syncFromPointerChanges(bootstrappingServersFromPointerChange, { ...options, fromTimestamp, pointerChangesWaitTime: 0 }, () => false)
+          syncingServers.add(bootstrappingServersFromPointerChange)
+          bootstrappingServersFromPointerChanges.delete(bootstrappingServersFromPointerChange)
+        } catch (error) {
+          // If there's an error, the server doesn't pass to syncing state
```
Comment (agusaldasoro) [inline src/synchronizer.ts:112]: Should then throw an error?

File: src/synchronizer.ts [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        return exponentialFallofRetryComponent
+      }
+    }
+  )
+
+  function createSyncJob() {
+    return createExponentialFallofRetry(logger, {
+      async action() {
+        try {
+          // metrics start?
```
Comment (agusaldasoro) [inline src/synchronizer.ts:166]: is this pending?

File: test/createCatalystDeploymentStream.spec.ts [hunk 13]
```diff
@@ -1,176 +1,180 @@
...
-      { entityType: 'profile', entityId: 'Qm000012', localTimestamp: 12, authChain, pointers: ['0x1'] },
-      { entityType: 'profile', entityId: 'Qm000013', localTimestamp: 13, authChain, pointers: ['0x1'] },
-    ])
-
-    expect(stream.getGreatesProcessedTimestamp()).toEqual(13)
-  })
-})
+
+it('test', () => { })
+// import { createCatalystDeploymentStream } from '../src'
```
Comment (agusaldasoro) [inline test/createCatalystDeploymentStream.spec.ts:3]: Would it be better to delete this test?

File: test/processed-snapshots.spec.ts [hunk 14]
```diff
@@ -31,47 +31,47 @@ describe('processed snapshots', () => {
...
-      processedSnapshots.endStreamOf(processedSnapshot, 1)
-      expect(await processedSnapshots.shouldStream(processedSnapshot)).toBeFalsy()
-    })
-
-    it('should return true when the snapshot and the replaced ones were not processed', async () => {
-      expect(await processedSnapshots.shouldStream(processedSnapshot, [h1, h2, h3])).toBeTruthy()
-    })
-
-  })
+  // describe('shouldStream', () => {
```
Comment (agusaldasoro) [inline test/processed-snapshots.spec.ts:34]: +1

File: src/client.ts [hunk 15]
```diff
@@ -1,22 +1,19 @@
...
+  Promise<Snapshot[]> {
   try {
     const incrementalSnapshotsUrl = new URL(`${server}/snapshots`).toString()
     // TODO: validate response
     const newSnapshots: {
       hash: string,
       timeRange: { initTimestamp: number, endTimestamp: number },
       replacedSnapshotHashes?: string[]
-    }[] = await components.downloadQueue.scheduleJobWithRetries(() => fetchJson(incrementalSnapshotsUrl, components.fetcher), 1)
+    }[] = await components.downloadQueue.scheduleJobWithRetries(() => fetchJson(incrementalSnapshotsUrl, components.fetcher), 3)
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/client.ts:16]: The idea is to remove the old `/snapshot` fetch in the future and use the parameter `retries`

File: src/exponential-fallof-retry.ts [hunk 16]
```diff
@@ -53,6 +56,10 @@ export function createExponentialFallofRetry(
 
       try {
         await options.action()
+        if (exitOnSuccess) {
+          logs.info('Breaking iteration. Action ended successfully')
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/exponential-fallof-retry.ts:60]: You can configure the component to run indefinitely with an interval no matter if the action finished successfully or not.
Or you can configure it to run "exponentially" failing until success. This is the case

File: src/file-processor.ts (hunk 2 above)
Comment (pedrotambo) [reply src/file-processor.ts:57]: No, that's where the errors come from in the types exported by `common-schemas`.

File: src/synchronizer.ts (hunk 11 above)
Comment (pedrotambo) [reply src/synchronizer.ts:112]: No, it runs bootstrap from pointer-changes for each catalyst. If one fails, it doesn't change its state from `bootstrapFromPointerChanges` to `syncingServers`. It'll be retried in the future

File: src/synchronizer.ts (hunk 12 above)
Comment (pedrotambo) [reply src/synchronizer.ts:166]: yes

File: test/createCatalystDeploymentStream.spec.ts (hunk 13 above)
Comment (pedrotambo) [reply test/createCatalystDeploymentStream.spec.ts:3]: yes, chore stuff and unit/integration testing will be done once the functionality is verified so this tasks are not done multiple times :(

File: test/processed-snapshots.spec.ts (hunk 14 above)
Comment (pedrotambo) [reply test/processed-snapshots.spec.ts:34]: +1

File: src/client.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (agusaldasoro) [reply src/client.ts:15]: can we use retries here instead of 3?

File: src/client.ts [hunk 17]
```diff
@@ -38,15 +58,25 @@ export async function* fetchJsonPaginated<T>(
   }
 }
 
-export function fetchPointerChanges(
+export async function* fetchPointerChanges(
```
Comment (agusaldasoro) [inline src/client.ts:61]: is this a breaking change of the lib? If that's the case remember to upgrade the major version when releasing

File: src/file-processor.ts (hunk 1 above)
Comment (agusaldasoro) [reply src/file-processor.ts:54]: what do you think of this suggestion?

File: src/index.ts [hunk 18]
```diff
@@ -123,92 +153,100 @@ export async function downloadEntityAndContentFiles(
...
-    options.contentServer,
-    options.requestMaxRetries
-  )
-
-  // 2. download the snapshot file if it contains deployments
-  //    in the range we are interested (>= genesisTimestamp)
-  if (hash && lastIncludedDeploymentTimestamp && lastIncludedDeploymentTimestamp > genesisTimestamp) {
+  if (shouldStreamSnapshot) {
     try {
       // 2.1. download the snapshot file if needed
```
Comment (agusaldasoro) [inline src/index.ts:177]: check the comments here, as there is no `// 1.` now

File: src/index.ts (hunk 9 above)
Comment (agusaldasoro) [reply src/index.ts:234]: check this comment

File: src/client.ts [hunk 19]
```diff
@@ -1,22 +1,19 @@
...
 export async function getSnapshots(components: SnapshotsFetcherComponents, server: string, retries: number):
-  Promise<{
-    hash: string,
-    lastIncludedDeploymentTimestamp: number
-    replacedSnapshotHashes?: string[]
-  }[]> {
+  Promise<Snapshot[]> {
   try {
     const incrementalSnapshotsUrl = new URL(`${server}/snapshots`).toString()
     // TODO: validate response
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/client.ts]: removed

File: src/processed-snapshots.ts [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+  async function saveIfStreamEndedAndAllEntitiesWereProcessed(snapshotHash: string) {
+    const numberOfEntities = numberOfProcessedEntitiesBySnapshot.get(snapshotHash)
+    if (snapshotsCompletelyStreamed.has(snapshotHash) && numberOfEntities == 0) {
+      await components.processedSnapshotStorage.saveProcessed(snapshotHash)
+      components.metrics.increment('dcl_processed_snapshots_total', { state: 'saved' })
+    }
+  }
+
+  return {
+    async shouldProcessSnapshot(snapshotHash: string, snapshotReplacedGroups: string[][]) {
```
Comment (agusaldasoro) [inline src/processed-snapshots.ts:38]: can we rename this to another that explains that it has a side effect?

File: test/createCatalystDeploymentStream.spec.ts (hunk 13 above)
Comment (pedrotambo) [reply test/createCatalystDeploymentStream.spec.ts]: file removed.

File: src/client.ts (hunk 17 above)
Comment (pedrotambo) [reply src/client.ts:61]: Yes, will be major version

File: src/index.ts (hunk 18 above)
Comment (pedrotambo) [reply src/index.ts:177]: updated

File: src/client.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/client.ts:15]: changed

File: src/synchronizer.ts [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+              ...bootstrappingServersFromSnapshots,
+              ...bootstrappingServersFromPointerChanges])
+            }`)
+        }
+        syncFinished.resolve()
+        // } catch (e: any) {
+        //   // we don't log the exception here because createExponentialFallofRetry(logger, options) receives the logger
+        //   // increment metrics
+        //   throw e
+        // }
```
Comment (agusaldasoro) [inline src/synchronizer.ts:263]:
```suggestion

```

File: src/synchronizer.ts [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+          await cb()
+        }
+      },
+      async onSyncFinished() {
+        await syncFinished
+      }
+    }
+  }
+
+  function removeServersNotToSyncFromStateSet(serversToSync: Set<string>, syncStateSet: Set<string>) {
```
Comment (agusaldasoro) [inline src/synchronizer.ts:285]:
```suggestion
  // Remove a server from sync list when it was removed from the Servers DAO
  function removeServersNotToSyncFromStateSet(serversToSync: Set<string>, syncStateSet: Set<string>) {
```

File: src/types.ts [hunk 23]
```diff
@@ -44,7 +50,12 @@ export type SnapshotsFetcherComponents = {
  * @public
  */
 export type IDeployerComponent = {
-  deployEntity(entity: DeploymentWithAuthChain, contentServers: string[]): Promise<void>
+  /**
+   * awaitng deployEntity might not imply that the entity was deployed. To be marked the entity as deployed, it needs
```
Comment (agusaldasoro) [inline src/types.ts:54]:
```suggestion
   * awaiting deployEntity might not imply that the entity was deployed. To be marked the entity as deployed, it needs
```

File: test/createCatalystDeploymentStream.spec.ts (hunk 13 above)
Comment (agusaldasoro) [reply test/createCatalystDeploymentStream.spec.ts]: Don't forget this! :)

File: package.json [hunk 24]
```diff
@@ -7,17 +7,17 @@
...
   },
   "keywords": [],
   "author": "Decentraland Contributors",
   "license": "Apache-2.0",
   "dependencies": {
-    "@dcl/catalyst-storage": "^1.0.0-20220604235145.commit-892b669",
+    "@dcl/catalyst-storage": "^2.0.0",
     "@dcl/hashing": "^1.0.0",
-    "@dcl/schemas": "^5.4.2",
+    "@dcl/schemas": "tag-fast-bootstrapping-5",
```
Comment (agusaldasoro) [inline package.json:20]: remember to change this!

File: src/synchronizer.ts (hunk 21 above)
Comment (pedrotambo) [reply src/synchronizer.ts]: removed

File: src/processed-snapshots.ts (hunk 20 above)
Comment (pedrotambo) [reply src/processed-snapshots.ts]: solved

**Review by agusaldasoro — APPROVED** [review]

**Pull request merged**